[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57288 [RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided**

If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

Also add pybind of `_get_device_map`.

The changes in unit test setup is separated out as a follow-up PR, as currently it breaks some tests in distributed/rpc/test_faulty_agent.py.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

Commit:
[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided

If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670


It seems that TensorPipe backend is much slower, and it causes test timeout if run `buck test @mode/dev-nosan caffe2/test/distributed/rpc:process_group_agent -- RemoteModule`, although the individual test can pass.

    ✗ Fail: caffe2/test/distributed/rpc:process_group_agent - test_unsupported_methods (fb.test_process_group_agent.ProcessGroupRemoteModuleTestWithFork) (124.102)
Test output:
> Repro command : $(cat "/tmp/tpx-20210428-165139.216124/6fc0ceaa-966b-49dd-a39f-afbea0e3b632/execution_command")
RuntimeError: Process 1 terminated or timed out after 100.16515851020813 seconds
  File "/usr/local/fbcode/platform009/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/local/fbcode/platform009/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/local/fbcode/platform009/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 374, in wrapper
    self._join_processes(fn)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 566, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 614, in _check_return_codes
    raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time))
stdout:
Could not download https://raw.githubusercontent.com/pytorch/test-infra/master/stats/slow-tests.json because of error <urlopen error [Errno 101] Network is unreachable>.
Couldn't download test skip set, leaving all tests enabled...
Timing out after 100 seconds and killing subprocesses.


Failed tests:

    ✗ caffe2/test/distributed/rpc:process_group_agent - test_forward_sync (test_process_group_agent.ProcessGroupRemoteModuleTestWithSpawn)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_bad_module (test_process_group_agent.ProcessGroupRemoteModuleTestWithSpawn)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_forward_async (test_process_group_agent.ProcessGroupRemoteModuleTestWithSpawn)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_remote_parameters (test_process_group_agent.ProcessGroupRemoteModuleTestWithSpawn)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_get_module_rref (test_process_group_agent.ProcessGroupRemoteModuleTestWithSpawn)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_forward_async_script (test_process_group_agent.ProcessGroupRemoteModuleTestWithSpawn)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_remote_parameters (fb.test_process_group_agent.ProcessGroupRemoteModuleTestWithFork)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_forward_async (fb.test_process_group_agent.ProcessGroupRemoteModuleTestWithFork)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_unsupported_methods (fb.test_process_group_agent.ProcessGroupRemoteModuleTestWithFork)
    ✗ caffe2/test/distributed/rpc:process_group_agent - test_forward_with_kwargs (test_process_group_agent.ProcessGroupRemoteModuleTestWithSpawn)

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -17,47 +17,46 @@
 # will not be implicitly moved to the same cuda device.
 def _remote_forward(
-    module_rref: RRef[module_interface_cls], device: str, {arg_types}){arrow_and_return_type}:
+    module_rref: RRef[module_interface_cls], device: str, is_device_map_set: bool, {arg_types}){arrow_and_return_type}:
```
Is `is_device_map_set` argument user-facing?

SciPioneer replied:
Not user-facing. See L70. We pass a wrapper of (internal args, user-facing args) into `_remote_forward`.

Commit:
Update on "[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided"


If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
(Can be done in a follow up PR)
Can the docs be enhanced to mention that inputs are automatically moved to the right device on the remote worker, and , for TensorPipe backend, GPU tensors are returned over the wire according to the device_maps set?

Commit:
Update on "[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided"


If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

The changes in unit test setup is separated out as a follow-up PR, as currently it breaks some tests in distributed/rpc/test_faulty_agent.py.

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

[ghstack-poisoned]

SciPioneer commented:
> inputs are automatically moved to the right device on the remote worker, and , for TensorPipe backend, GPU tensors are returned over the wire according to the device_maps set?

Thanks! Added some descriptions to RemoteModule class.

Commit:
Update on "[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided"


If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

The changes in unit test setup is separated out as a follow-up PR, as currently it breaks some tests in distributed/rpc/test_faulty_agent.py.

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided"


If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

The changes in unit test setup is separated out as a follow-up PR, as currently it breaks some tests in distributed/rpc/test_faulty_agent.py.

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided"


If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

Also add pybind of `_get_device_map`.

The changes in unit test setup is separated out as a follow-up PR, as currently it breaks some tests in distributed/rpc/test_faulty_agent.py.

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided"


If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

Also add pybind of `_get_device_map`.

The changes in unit test setup is separated out as a follow-up PR, as currently it breaks some tests in distributed/rpc/test_faulty_agent.py.

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided"


If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

Also add pybind of `_get_device_map`.

The changes in unit test setup is separated out as a follow-up PR, as currently it breaks some tests in distributed/rpc/test_faulty_agent.py.

Still need to fix test_load_di_parts in `torch/fb/training_toolkit/applications/sparse_nn/batch_distributed_inference/tests:batch_distributed_inference_test`. Currently an early return is used to bypass this test failure.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

[ghstack-poisoned]
